includes/gateways/paypal/includes/class-wc-gateway-paypal-request.php

```diff
@@ -180,61 +236,87 @@ protected function get_shipping_args( $order ) {
 		return $shipping_args;
 	}
 
+	/**
+	 * Get shipping cost line item args for paypal request.
+	 *
+	 * @param  WC_Order $order Order object.
+	 * @param  bool     $include_shipping_tax Whether to include shipping tax or not.
+	 * @return array
+	 */
+	protected function get_shipping_cost_line_item( $order, $include_shipping_tax ) {
+		$line_item_args = array();
+		$shipping_total = $order->get_shipping_total();
+		if ( $include_shipping_tax ) {
+			$shipping_total += $order->get_shipping_tax();
+		}
+
+		// Add shipping costs. Paypal ignores anything over 5 digits (999.99 is the max).
+		// We also check that shipping is not the **only** cost as PayPal won't allow payment
+		// if the items have no cost.
+		if ( $order->get_shipping_total() > 0 && $order->get_shipping_total() < 999.99 && $this->number_format( $order->get_shipping_total() + $order->get_shipping_tax(), $order ) !== $this->number_format( $order->get_total(), $order ) ) {
+			$line_item_args['shipping_1'] = $this->number_format( $shipping_total, $order );
+		} elseif ( $order->get_shipping_total() > 0 ) {
+			/* translators: %s: Order shipping method */
+			$this->add_line_item( sprintf( __( 'Shipping via %s', 'woocommerce' ), $order->get_shipping_method() ), 1, $this->number_format( $shipping_total, $order ) );
+		}
+
+		return $line_item_args;
+	}
+
+	/**
+	 * Get line item args for paypal request as a single line item.
+	 *
+	 * @param  WC_Order $order Order object.
+	 * @param  bool     $include_shipping_tax Whether to include shipping tax or not.
+	 * @return array
+	 */
+	protected function get_line_item_args_single_item( $order, $include_shipping_tax ) {
+		$this->delete_line_items();
+
+		$all_items_name = $this->get_order_item_names( $order );
+		$this->add_line_item( $all_items_name ? $all_items_name : __( 'Order', 'woocommerce' ), 1, $this->number_format( $order->get_total() - $this->round( $order->get_shipping_total() + $order->get_shipping_tax(), $order ), $order ), $order->get_order_number() );
+		$line_item_args = $this->get_shipping_cost_line_item( $order, $include_shipping_tax );
+
+		return array_merge( $line_item_args, $this->get_line_items() );
+	}
+
 	/**
 	 * Get line item args for paypal request.
 	 *
 	 * @param  WC_Order $order Order object.
+	 * @param  bool     $force_one_line_item Create only one item for this order.
 	 * @return array
 	 */
-	protected function get_line_item_args( $order ) {
+	protected function get_line_item_args( $order, $force_one_line_item = false ) {
 
 		/**
 		 * Try passing a line item per product if supported.
 		 */
 		if ( ( ! wc_tax_enabled() || ! wc_prices_include_tax() ) && $this->prepare_line_items( $order ) ) {
+			$line_item_args       = array();
+			$include_shipping_tax = false;
+			if ( $force_one_line_item ) {
```

Comment:
Seems really odd to do this here and then repeat the single item code lower down. Why can't we check this on line 295?

Reply:
..because the `$include_shipping_tax` is different when generating single line item request in this branch and in the else branch for single line item.

The logic gets a bit convoluted here, but I understand it in the following way:
If we add `! $force_one_line_item` check to line 295, then we would need to check for `( ! wc_tax_enabled() || ! wc_prices_include_tax() ) && $this->prepare_line_items( $order )` again in the single line item branch to see if the shipping tax should be included, otherwise we would be changing the behaviour. 

I'm not 100% sure if this logic is correct (that inclusion of shipping tax depends on `wc_tax_enabled()`, `wc_prices_include_tax()` and also `prepare_line_items()`), maybe you can help me to find out/point me in the right direction, but I didn't want to change the logic.

To improve the code, I extracted the `include shipping tax` flag before the larger if statement, maybe it looks a bit better now?